### PR TITLE
[Refactor] Remove _run_evaluation + 3 EvaluationLoop

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR creates 2 validation Loop. One in FitLoop and one on the trainer.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py [hunk 1]
```diff
@@ -99,6 +101,23 @@ def reset(self) -> None:
         if isinstance(self._max_batches, int):
             self._max_batches = [self._max_batches] * len(self.dataloaders)
 
+    def run(self, *args: Any, **kwargs: Any) -> Optional[Any]:
+        if not (self.trainer.evaluating or self.trainer.sanity_checking):
+            rank_zero_warn(
+                f"`validation` was called but the running stage is set to {self.trainer.state.stage}."
+                " This should not happen normally. Setting it to `RunningStage.VALIDATING`", RuntimeWarning
+            )
+            self.validating = True
```
Comment (carmocca) [inline pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py:110]: I added this back when I did the runningstage refactor just in case, but nothing has come up

Maybe we could change it now to an `assert`

File: pytorch_lightning/trainer/properties.py [hunk 2]
```diff
@@ -488,6 +489,14 @@ def train_loop(self) -> FitLoop:
         # FIXME(@awaelchli): the current train_loop should be renamed to fit_loop
         return self.fit_loop
 
+    @property
+    def evaluation_loop(self) -> EvaluationDataLoaderLoop:
+        if self.training and self.sanity_checking:
```
Comment (carmocca) [inline pytorch_lightning/trainer/properties.py:494]: This won't ever return True. You meant

`self.trainer.state.fn == TrainerFn.FIT and self.sanity_checking`

but no need for the first condition as sanity_checking is only done for fit

```suggestion
        if self.sanity_checking:
```

File: pytorch_lightning/trainer/properties.py [hunk 3]
```diff
@@ -488,6 +489,14 @@ def train_loop(self) -> FitLoop:
...
+            return self.fit_loop.validation_loop
+        elif self.validating:
+            return self.validation_loop
+        return self.test_loop
```
Comment (carmocca) [inline pytorch_lightning/trainer/properties.py:498]: elif for this one too? to avoid silent correctness

File: pytorch_lightning/trainer/properties.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/trainer/properties.py:494]: Thanks ! I wasn't sure.

File: pytorch_lightning/trainer/properties.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/trainer/properties.py:498]: Added a MisConfigurationException.

File: pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py [hunk 4]
```diff
@@ -99,6 +101,23 @@ def reset(self) -> None:
...
+
+        self.reload_evaluation_dataloaders()
+
+        with torch.no_grad():
+            # the model is set to eval mode in on_run_start() and back to train mode in on_run_end()
+            eval_loop_results = super().run(*args, **kwargs)
+
+        eval_loop_results = eval_loop_results or []
```
Comment (awaelchli) [inline pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py:118]: this can go into on_run_end then?

File: pytorch_lightning/trainer/properties.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/trainer/properties.py:498]: I'd say `RuntimeError` would be a better fit, as this would be an error on our side, not the user's

File: pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py [hunk 5]
```diff
@@ -99,6 +101,23 @@ def reset(self) -> None:
...
 
+    def run(self, *args: Any, **kwargs: Any) -> Optional[Any]:
+        if not (self.trainer.evaluating or self.trainer.sanity_checking):
+            rank_zero_warn(
+                f"`validation` was called but the running stage is set to {self.trainer.state.stage}."
+                " This should not happen normally. Setting it to `RunningStage.VALIDATING`", RuntimeWarning
+            )
+            self.validating = True
+
+        self.reload_evaluation_dataloaders()
```
Comment (awaelchli) [inline pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py:112]: note, self.reload_dataloaders has to go outside the run() otherwise the skip condition does not know max batches.

File: pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py [hunk 6]
```diff
@@ -99,6 +101,23 @@ def reset(self) -> None:
...
+
+        with torch.no_grad():
```
Comment (awaelchli) [inline pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py:114]: I vote for keeping this outside or adding a decorator.

File: pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py:118]: Good point !

File: pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py:114]: Adding out.

File: pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py:101]: Good point !

File: pytorch_lightning/trainer/properties.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/trainer/properties.py]: Done !

File: pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py]: Done !

File: pytorch_lightning/loops/fit_loop.py [hunk 7]
```diff
@@ -54,6 +55,7 @@ def __init__(
         self.max_epochs = 1000 if (max_epochs is None and max_steps is None) else max_epochs
         self.min_epochs = 1 if (min_epochs is None and min_steps is None) else min_epochs
         self.training_loop = TrainingEpochLoop(min_steps, max_steps)
+        self.validation_loop = EvaluationDataLoaderLoop()
```
Comment (carmocca) [inline pytorch_lightning/loops/fit_loop.py:58]: @awaelchli has it been decided that we do the full name instead of the shorter one?

```suggestion
        self.val_loop = EvaluationDataLoaderLoop()
```

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 8]
```diff
@@ -300,5 +300,6 @@ def progress_bar_metrics(self) -> Dict[str, float]:
 
     def teardown(self):
         self.trainer.train_loop.results.cpu()
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:302]: Make this one protected too?

File: pytorch_lightning/trainer/properties.py [hunk 9]
```diff
@@ -488,6 +489,16 @@ def train_loop(self) -> FitLoop:
         # FIXME(@awaelchli): the current train_loop should be renamed to fit_loop
         return self.fit_loop
 
+    @property
+    def evaluation_loop(self) -> EvaluationDataLoaderLoop:
+        if self.training or self.sanity_checking:
```
Comment (carmocca) [inline pytorch_lightning/trainer/properties.py:494]: Actually, we set `self.trainer.validating = True` before accessing this property so we do need:

```suggestion
        if self.state.fn == TrainerFn.FIT or self.sanity_checking:
```

Comment (tchaton) [reply pytorch_lightning/trainer/properties.py:494]: Good call !

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py]: Done !

File: pytorch_lightning/trainer/properties.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/trainer/properties.py]: Done !

File: pytorch_lightning/loops/training_epoch_loop.py [hunk 10]
```diff
@@ -151,7 +156,7 @@ def on_advance_end(self):
         should_check_val = self.should_check_val_fx(self.iteration_count, self.is_last_batch)
         if should_check_val:
             self.trainer.validating = True
-            self.trainer._run_evaluation()
+            self.trainer.fit_loop.validation_loop.run()
```
Comment (awaelchli) [inline pytorch_lightning/loops/training_epoch_loop.py:159]: careful, before, run_evaluation would 1) reset the val dataloader 2) torch.no_grad.

Comment (tchaton) [reply pytorch_lightning/loops/training_epoch_loop.py]: Done !

File: pytorch_lightning/loops/base.py [hunk 11]
```diff
@@ -61,6 +61,14 @@ def connect(self, trainer: 'pl.Trainer', *args: Any, **kwargs: Any) -> None:
         """Connects Loop with all the necessary things like connectors and accelerators."""
         self.trainer = proxy(trainer)
 
+    def on_skip(self) -> Optional[Any]:
+        """
+        The function to run when :attr:`run` should be skipped.
```
Comment (awaelchli) [inline pytorch_lightning/loops/base.py:66]:
```suggestion
        The function to run when :meth:`run` should be skipped, determined by the condition in :attr:`skip`.
```

File: pytorch_lightning/loops/base.py [hunk 12]
```diff
@@ -61,6 +61,14 @@ def connect(self, trainer: 'pl.Trainer', *args: Any, **kwargs: Any) -> None:
...
+
+        Returns:
+            the default output value of :attr:`on_run_end`
```
Comment (awaelchli) [inline pytorch_lightning/loops/base.py:69]:
```suggestion
            the default output value of :meth:`on_run_end`
```

File: pytorch_lightning/loops/base.py [hunk 13]
```diff
@@ -61,6 +61,14 @@ def connect(self, trainer: 'pl.Trainer', *args: Any, **kwargs: Any) -> None:
         """Connects Loop with all the necessary things like connectors and accelerators."""
         self.trainer = proxy(trainer)
 
+    def on_skip(self) -> Optional[Any]:
```
Comment (awaelchli) [inline pytorch_lightning/loops/base.py:64]: I like it

File: pytorch_lightning/loops/fit_loop.py (hunk 7 above)
Comment (awaelchli) [reply pytorch_lightning/loops/fit_loop.py:58]: no but highest priority is consistency over anything else imo

File: pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py [hunk 14]
```diff
@@ -99,6 +94,14 @@ def reset(self) -> None:
         if isinstance(self._max_batches, int):
             self._max_batches = [self._max_batches] * len(self.dataloaders)
 
+    def on_skip(self):
+        return []
+
+    def run(self, *args: Any, **kwargs: Any) -> Optional[Any]:
+        assert self.trainer.evaluating or self.trainer.sanity_checking
+        # the model is set to eval mode in `on_run_start()` and back to train mode in `on_run_end()`
+        return super().run(*args, **kwargs)
```
Comment (awaelchli) [inline pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py:103]: can we just remove it? 
We also dont have assert trainer.training in fit loop etc
cc @carmocca

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/loops/fit_loop.py (hunk 7 above)
Comment (carmocca) [reply pytorch_lightning/loops/fit_loop.py:58]: Is that an argument in favor of the long or short name?

File: pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py (hunk 14 above)
Comment (carmocca) [reply pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py:103]: We can now but could be helpful when we allow people to call run at free will

File: pytorch_lightning/loops/base.py (hunk 11 above)
Comment (carmocca) [reply pytorch_lightning/loops/base.py:66]: Note that you used ``` :attr:`run` ``` in other docstrings. That's another TODO for you 😂

File: pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py (hunk 14 above)
Comment (carmocca) [reply pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py:103]:
```suggestion
```

File: pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py [hunk 15]
```diff
@@ -99,6 +94,10 @@ def reset(self) -> None:
         if isinstance(self._max_batches, int):
             self._max_batches = [self._max_batches] * len(self.dataloaders)
 
+    def on_skip(self):
+        return []
+
```
Comment (carmocca) [inline pytorch_lightning/loops/dataloader/evaluation_dataloader_loop.py:99]:
```suggestion
    def on_skip(self) -> List:
        return []
```

File: pytorch_lightning/loops/base.py (hunk 11 above)
Comment (awaelchli) [reply pytorch_lightning/loops/base.py]: where did I do that?

Comment (carmocca) [reply pytorch_lightning/loops/base.py]:
```python
pytorch_lightning/loops//base.py:29:        * :attr:`reset` (method): Resets the internal state between multiple calls of :attr:`run`
pytorch_lightning/loops//base.py:30:        * :attr:`advance` (method): Implements one step of the loop
pytorch_lightning/loops//base.py:68:        Will frequently check the :attr:`done` condition and calls :attr:`advance`
pytorch_lightning/loops//base.py:95:        """Resets the internal state of the loop at the beginning of each call to :attr:`run`."""
pytorch_lightning/loops//base.py:99:        Hook to be called as the first thing after entering :attr:`run` (except the state reset).
pytorch_lightning/loops//base.py:101:        Accepts all arguments passed to :attr:`run`.
pytorch_lightning/loops//base.py:107:        Hook to be called each time before :attr:`advance` is called. Accepts all arguments passed to :attr`run`.
pytorch_lightning/loops//base.py:113:        """Performs a single step. Accepts all arguments passed to :attr:`run`."""
pytorch_lightning/loops//base.py:116:        """Hook to be called each time after :attr:`advance` is called."""
pytorch_lightning/loops//base.py:119:        """Hook to be called at the end of the run. Its return argument is returned from :attr:`run`."""
```

Should all these be `:meth:`?

Comment (awaelchli) [reply pytorch_lightning/loops/base.py]: yes, meth is for methods and attr for properties. They also need the class normally. 
but doesn't matter, docs are not visible yet.

**Review by justusschock — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/loops/base.py [hunk 16]
```diff
@@ -69,10 +77,10 @@ def run(self, *args: Any, **kwargs: Any) -> Optional[Any]:
         until :attr:`done` evaluates to ``True``.
 
         Returns:
-            the output of :attr`on_run_end` (often outputs collected from each step of the loop)
+            the output of :attr:`on_run_end` (often outputs collected from each step of the loop)
         """
         if self.skip:
-            return
+            return self.on_skip()
```
Comment (ananthsub) [inline pytorch_lightning/loops/base.py:83]: in this case, why does the caller call `run` if skip is publicly exposed?
